src/Stratis.Features.FederatedPeg/Wallet/FederationWalletManager.cs

```diff
@@ -148,6 +153,87 @@ public class FederationWalletManager : LockProtected, IFederationWalletManager
+
+                if (!spendTxsByBlockId.TryGetValue(spendingDetail.BlockHash, out Dictionary<uint256, List<TransactionData>> spentOutputsBySpendTxId))
+                {
+                    spentOutputsBySpendTxId = new Dictionary<uint256, List<TransactionData>>();
+                    spendTxsByBlockId[spendingDetail.BlockHash] = spentOutputsBySpendTxId;
+                }
+
+                if (!spentOutputsBySpendTxId.TryGetValue(spendingDetail.TransactionId, out List<TransactionData> spentOutputs))
+                {
+                    spentOutputs = new List<TransactionData>();
+                    spentOutputsBySpendTxId[spendingDetail.TransactionId] = spentOutputs;
+                }
+
+                spentOutputs.Add(transactionData);
+            }
+
+            // Will keep track of the height of spending details we're unable to fix.
+            int firstMissingTransactionHeight = this.LastBlockHeight() + 1;
+
+            // Try to fix the spending details transaction Hex values.
+            foreach ((uint256 blockId, Dictionary<uint256, List<TransactionData>> spentOutputsBySpendTxId) in spendTxsByBlockId)
+            {
+                Block block = this.blockStore.GetBlock(blockId);
+
+                foreach ((uint256 spendTxId, List<TransactionData> spentOutputs) in spentOutputsBySpendTxId)
+                {
+                    Transaction spendTransaction = block.Transactions.FirstOrDefault(t => t.GetHash() == spendTxId);
+
+                    if (spendTransaction != null)
+                    {
+                        string hex = spendTransaction.ToHex();
+
+                        foreach (TransactionData transactionData in spentOutputs)
+                            transactionData.SpendingDetails.Hex = hex;
+                    }
+                    else
+                    {
+                        // The spending transaction could not be found in the consensus chain.
+                        // Set the maxValidHeight to the block before where the spending transaction occurred.
+                        foreach (TransactionData transactionData in spentOutputs)
+                        {
+                            SpendingDetails spendingDetails = transactionData.SpendingDetails;
+
+                            Guard.Assert(spendingDetails.BlockHeight != null);
+
+                            if (spendingDetails.BlockHeight < firstMissingTransactionHeight)
+                                firstMissingTransactionHeight = (int)spendingDetails.BlockHeight;
+                        }
+                    }
+                }
+            }
+
+            // If there are unresolvable spending details then re-sync from that point onwards.
+            if (firstMissingTransactionHeight <= this.LastBlockHeight())
+            {
+                ChainedHeader fork = this.chainIndexer.GetHeader(Math.Min(firstMissingTransactionHeight - 1, this.chainIndexer.Height));
+
+                this.RemoveBlocks(fork);
```

Comment:
Should we immediately force processing here of any blocks that are between the forkpoint and the tip? Otherwise it looks like we'd have to wait for some other event (i.e. a new block arriving) to trigger the processing.

Reply:
I think `RemoveBlocks` does all the processing we require at this point unless I'm missing something.